File: go/remote.go
package aight

import (
	"bytes"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"
)

// DefaultIngestURL is where Push sends data when no url is given and
// AIGHT_INGEST_URL isn't set.
const DefaultIngestURL = "https://api.aight.studio/api/ingest/spans"

// spanFrame is one frame of a row's call chain. The snippet is the source
// window captured around that line, when it could be read — omitted otherwise,
// which the Workspace shows as "not captured".
type spanFrame struct {
	Filepath string `json:"filepath"`
	Lineno   int    `json:"lineno"`
	Function string `json:"function"`
	// 1-indexed line the snippet text starts at; only meaningful with Snippet.
	SnippetStart int    `json:"snippet_start,omitempty"`
	Snippet      string `json:"snippet,omitempty"`
}

// frameFrom builds one chain frame, attaching the captured snippet if there is
// one — mirrors _frame_dict in the Python SDK's remote.py.
func frameFrom(filepath string, lineno int, function string) spanFrame {
	f := spanFrame{Filepath: filepath, Lineno: lineno, Function: function}
	if s, ok := snippetFor(filepath, lineno); ok {
		f.SnippetStart, f.Snippet = s.Start, s.Text
	}
	return f
}

// spanRow is one ingest row: a call chain plus the usage attributed to its
// leaf. The server requires "chain" — a flat (filepath, lineno, function)
// row is rejected with "missing 'chain'" (see README.md "Wire
// protocol"). This SDK attributes a call to exactly one line, so that line
// is sent as a one-frame chain; real multi-frame chains need stack walking
// here, the way the Python SDK does it.
//
// Model and the four token counts go with it, and nothing else: the server
// prices the row from its own table at receive time, and an absent cost is how
// this SDK says "making no claim" rather than "$0.00". Sending a cost from here
// would be a second copy of that table, and a row naming a model the SDK knew
// and the server didn't was priced locally and never repriced.
type spanRow struct {
	Chain        []spanFrame `json:"chain"`
	Calls        int         `json:"calls"`
	Model        string      `json:"model"`
	InputTokens  int         `json:"input_tokens"`
	OutputTokens int         `json:"output_tokens"`
	// Their own counts, never folded into InputTokens. The platform sums four
	// rates and freezes the result at receive, so an input count that already
	// contained the cached tokens would bill them twice, permanently.
	CacheReadTokens     int `json:"cache_read_tokens"`
	CacheCreationTokens int `json:"cache_creation_tokens"`
	// Summed over the bucket's calls; the server divides by `calls`. Absent
	// rather than 0 when the caller never timed the call, so the Workspace
	// shows "—" instead of a real 0ms.
	LatencyMs float64 `json:"latency_ms,omitempty"`
	// Empty when the caller never set one, in which case the backend matches
	// the spend to business events by time window instead of explicitly.
	TraceID string `json:"trace_id"`
}

// Push sends the processor's current buckets to your hosted AIght Workspace.
// apiKey and url may be empty, in which case AIGHT_API_KEY and
// AIGHT_INGEST_URL (or DefaultIngestURL) are used.
func Push(p *CostByLineProcessor, apiKey, url string) error {
	if apiKey == "" {
		apiKey = os.Getenv("AIGHT_API_KEY")
	}
	if apiKey == "" {
		return fmt.Errorf("no API key: pass one, or set AIGHT_API_KEY (get one from the Integrate tab of your aight.studio dashboard)")
	}
	if url == "" {
		url = os.Getenv("AIGHT_INGEST_URL")
	}
	if url == "" {
		url = DefaultIngestURL
	}

	snap := p.Snapshot()
	if len(snap) == 0 {
		return nil
	}
	rows := make([]spanRow, 0, len(snap))
	for k, b := range snap {
		rows = append(rows, spanRow{
			Chain:        []spanFrame{frameFrom(k.Filepath, k.Lineno, k.Function)},
			Calls:        b.Calls,
			Model:        k.Model,
			InputTokens:  b.InputTokens,
			OutputTokens: b.OutputTokens,

			CacheReadTokens:     b.CacheReadTokens,
			CacheCreationTokens: b.CacheCreationTokens,
			LatencyMs:           b.LatencyMs,
			TraceID:             b.TraceID,
		})
	}

	body, err := json.Marshal(rows)
	if err != nil {
		return err
	}
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("X-Aight-Sdk-Language", "go")
	req.Header.Set("X-Aight-Sdk-Version", SDKVersion)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return fmt.Errorf("aight push failed: %s", resp.Status)
	}
	// Only after the server has acknowledged. The ingest endpoint *adds* each
	// row to what it already holds, so re-sending these running totals on a
	// later push would count every call a second time and quietly inflate
	// the AIght Workspace's spend. Clearing on success makes a second push
	// send only what happened since the first.
	//
	// ponytail: no idempotency key, so a push whose response is lost still
	// double-counts on retry. Add a client run-id if that ever bites.
	p.Clear()
	return nil
}

type valueRow struct {
	Filepath string  `json:"filepath"`
	ValueUSD float64 `json:"value_usd"`
}

// PushValue sends the current ValueSnapshot to /api/ingest/value. Same auth
// and defaulting rules as Push — including the clear-on-success, for the same
// reason: ReportValue only ever *adds* to the value it hands you, so a second
// push of the same running totals would insert the same earnings again.
// /api/ingest/value is a plain INSERT with no dedup, so that inflation would
// be permanent.
func PushValue(apiKey, url string) error {
	if apiKey == "" {
		apiKey = os.Getenv("AIGHT_API_KEY")
	}
	if apiKey == "" {
		return fmt.Errorf("no API key: pass one, or set AIGHT_API_KEY (get one from the Integrate tab of your aight.studio dashboard)")
	}
	if url == "" {
		url = os.Getenv("AIGHT_VALUE_INGEST_URL")
	}
	if url == "" {
		url = DefaultIngestURL[:len(DefaultIngestURL)-len("spans")] + "value"
	}

	snap := ValueSnapshot()
	if len(snap) == 0 {
		return nil
	}
	rows := make([]valueRow, 0, len(snap))
	for filepath, value := range snap {
		rows = append(rows, valueRow{Filepath: filepath, ValueUSD: value})
	}

	body, err := json.Marshal(rows)
	if err != nil {
		return err
	}
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("X-Aight-Sdk-Language", "go")
	req.Header.Set("X-Aight-Sdk-Version", SDKVersion)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return fmt.Errorf("aight push failed: %s", resp.Status)
	}
	// Only after the server has acknowledged. A failed push keeps the total, so
	// a retry can still send it.
	ClearValue()
	return nil
}

// eventRow is one business event. A single object on the wire, not an array —
// /api/ingest/events is the one ingest endpoint shaped that way.
type eventRow struct {
	EventID   string  `json:"event_id"`
	EventName string  `json:"event_name"`
	Value     float64 `json:"value"`
	Currency  string  `json:"currency"`
	// Unix seconds, fractional — not milliseconds.
	Timestamp float64 `json:"timestamp"`
	TraceID   string  `json:"trace_id"`
}

// EventOptions carries the optional half of PushEvent. Every zero value means
// "use the default": a generated event id, USD, now, no trace id, and the
// default events endpoint (AIGHT_EVENTS_INGEST_URL / AIGHT_API_KEY).
type EventOptions struct {
	EventID  string
	TraceID  string
	Currency string
	APIKey   string
	URL      string
	// Timestamp is Unix seconds. Zero means now.
	Timestamp float64
}

// PushEvent reports a business KPI — a Stripe checkout, a signup, a support
// ticket closed, whatever the caller's business counts — to power the AIght
// Workspace's fleet view and ROI attribution.
//
// Set EventOptions.TraceID to the same id passed to TracedLLMCallInfo for the
// run that produced the event; that is what gives the backend an EXPLICIT
// match instead of its time-window fallback.
//
// Returns the ingest endpoint's decoded JSON response, shaped
// {"ingested": n, "results": [...]} — each entry of "results" carries the
// match_type it resolved to (EXPLICIT/IMPLICIT/UNMATCHED). It is not the match
// result itself.
func PushEvent(eventName string, value float64, opts EventOptions) (map[string]any, error) {
	apiKey := opts.APIKey
	if apiKey == "" {
		apiKey = os.Getenv("AIGHT_API_KEY")
	}
	if apiKey == "" {
		return nil, fmt.Errorf("no API key: pass one, or set AIGHT_API_KEY (get one from the Integrate tab of your aight.studio dashboard)")
	}
	url := opts.URL
	if url == "" {
		url = os.Getenv("AIGHT_EVENTS_INGEST_URL")
	}
	if url == "" {
		url = strings.Replace(DefaultIngestURL, "/spans", "/events", 1)
	}

	eventID := opts.EventID
	if eventID == "" {
		var err error
		if eventID, err = newEventID(); err != nil {
			return nil, err
		}
	}
	currency := opts.Currency
	if currency == "" {
		currency = "USD"
	}
	timestamp := opts.Timestamp
	if timestamp == 0 {
		timestamp = float64(time.Now().UnixNano()) / 1e9
	}

	body, err := json.Marshal(eventRow{
		EventID:   eventID,
		EventName: eventName,
		Value:     value,
		Currency:  currency,
		Timestamp: timestamp,
		TraceID:   opts.TraceID,
	})
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("X-Aight-Sdk-Language", "go")
	req.Header.Set("X-Aight-Sdk-Version", SDKVersion)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("aight push failed: %s", resp.Status)
	}
	var out map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}

// newEventID returns a random RFC 4122 v4 UUID. Hand-rolled rather than pulled
// in: this SDK is stdlib-only, and 16 random bytes plus two version bits is
// the whole algorithm.
func newEventID() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", fmt.Errorf("aight: could not generate an event id: %w", err)
	}
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // RFC 4122 variant
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16]), nil
}

File: node/src/remote.js
// Push locally-traced spend data to your hosted AIght Workspace.
// Uses the built-in fetch (Node >=18) — no HTTP client dependency.
import { randomUUID } from 'node:crypto'

import { snippetFor } from './tracing.js'

export const DEFAULT_INGEST_URL = 'https://api.aight.studio/api/ingest/spans'

// Must match projects.CURRENT_SDK_VERSIONS["node"] on the backend — bump
// both together on release.
export const SDK_VERSION = '0.1.0'

function requireKey(apiKey) {
  const key = apiKey || process.env.AIGHT_API_KEY
  if (!key) {
    throw new Error(
      'No API key. Pass { apiKey }, or set AIGHT_API_KEY (get one from the Integrate tab of your aight.studio dashboard).'
    )
  }
  return key
}

/**
 * POST a JSON body and return the Response. All three ingest endpoints go
 * through here, so all three carry the same headers and the same timeout.
 *
 * On failure it keeps the server's own explanation. `res.ok === false` alone
 * says nothing: the ingest endpoints answer 401/400/402 with a `{"detail": ...}`
 * body saying *why*, and "Unknown or revoked API key" is the difference between
 * re-minting a key in five seconds and an afternoon of reading logs. Mirrors
 * aight.remote._send.
 */
async function send(url, key, payload) {
  const res = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${key}`,
      'X-Aight-Sdk-Language': 'node',
      'X-Aight-Sdk-Version': SDK_VERSION,
    },
    body: JSON.stringify(payload),
    // Matches the 10s the other three SDKs use. Without it a black-holing
    // endpoint (proxy, SG misconfig) leaves the agent's shutdown path pending
    // forever instead of failing with an error it can act on.
    signal: AbortSignal.timeout(10_000),
  })
  if (!res.ok) {
    const detail = await res
      .json()
      .then((body) => body?.detail)
      .catch(() => null)
    throw new Error(
      `aight push to ${url} failed: HTTP ${res.status} ${res.statusText}` +
        (detail ? ` — ${typeof detail === 'string' ? detail : JSON.stringify(detail)}` : '')
    )
  }
  return res
}

/** One chain frame, plus the captured source window when there is one.
 * Mirrors remote.py's _frame_dict. */
function frameDict(filepath, lineno, fn) {
  const frame = { filepath, lineno, function: fn }
  const snippet = snippetFor(filepath, lineno)
  if (snippet) {
    frame.snippet_start = snippet.start
    frame.snippet = snippet.text
  }
  return frame
}

export async function push(costProcessor, { apiKey, url } = {}) {
  const key = requireKey(apiKey)
  const endpoint = url || process.env.AIGHT_INGEST_URL || DEFAULT_INGEST_URL
  // Every row carries a full call chain (see README.md "Wire protocol").
  // This SDK attributes a call to exactly one line, so that line goes out as a
  // one-frame chain. Sending it flat instead is what the server rejects with
  // "missing 'chain'" — real multi-frame chains need stack walking here, the
  // way the Python SDK does it.
  //
  // No price on the row: this SDK computes none. The row names its model and
  // carries the token counts, and the server prices it from its own table.
  const rows = costProcessor.entries().map((e) => ({
    chain: [frameDict(e.filepath, e.lineno, e.function)],
    calls: e.calls,
    model: e.model,
    input_tokens: e.inputTokens,
    output_tokens: e.outputTokens,
    // Their own counts, never folded into input_tokens. The platform sums four
    // rates and freezes the result at receive, so an inclusive input count
    // would bill the cached tokens twice, permanently.
    cache_read_tokens: e.cacheReadTokens,
    cache_creation_tokens: e.cacheCreationTokens,
    // `undefined` drops the key from JSON.stringify entirely, which is what
    // this wants: an unmeasured call shows as "—" in the Workspace, where a
    // stored 0 would read as a real, impossibly fast measurement.
    latency_ms: e.latencyMs || undefined,
    // Ties this spend to any business event pushed with the same trace_id, so
    // the server matches the two EXPLICITLY instead of guessing by time window.
    // '' means nobody set one — not a match, rather than a wrong one.
    trace_id: e.traceId,
  }))
  if (rows.length === 0) return

  await send(endpoint, key, rows)
  // Only after the server has acknowledged. The ingest endpoint *adds* each row
  // to what it already holds, so re-sending these running totals on a later
  // push would count every call a second time and quietly inflate the
  // AIght Workspace's spend. Clearing on success makes a second push send
  // only what happened since the first.
  // ponytail: no idempotency key, so a push whose response is lost still
  // double-counts on retry. Add a client run-id if that ever bites.
  costProcessor.clear()
}

/**
 * Sends everything reportValue() recorded to /api/ingest/value. Same auth and
 * shape as push(), and the same clear-on-success for the same reason:
 * reportValue() only ever *adds* to the map it hands you, so a second push of
 * the same running totals would insert the same earnings again. The value
 * endpoint is a plain INSERT with no dedup, so that inflation would be
 * permanent.
 */
export async function pushValue(valueByFile, { apiKey, url } = {}) {
  const key = requireKey(apiKey)
  const endpoint = url || process.env.AIGHT_VALUE_INGEST_URL || DEFAULT_INGEST_URL.replace('/spans', '/value')
  const rows = [...valueByFile.entries()].map(([filepath, value]) => ({ filepath, value_usd: value }))
  if (rows.length === 0) return

  await send(endpoint, key, rows)
  // Only after the server has acknowledged. A failed push keeps the map, so a
  // retry can still send it.
  valueByFile.clear()
}

/**
 * Report a business KPI/event (a Stripe checkout, a signup, a support ticket
 * closing, ...) to power the AIght Workspace's fleet view and hybrid ROI
 * attribution. One event is one object: /api/ingest/events takes a single
 * object or a list, unlike the array-shaped /spans and /value endpoints.
 *
 * Pass `traceId` (the same one you gave tracedLlmCall) to tie this event to
 * that run as an EXPLICIT match. Without it the server falls back to its
 * time-window guess, which is the weakest data in the system.
 *
 * Returns the ingest endpoint's JSON response, {"ingested": n, "results": [...]}
 * — each result carries the match_type ("EXPLICIT"/"IMPLICIT"/"UNMATCHED") it
 * resolved to, plus its confidence_score. It is not the match result itself.
 */
export async function pushEvent(
  eventName,
  value = 0,
  { currency = 'USD', eventId, traceId = '', timestamp, apiKey, url } = {}
) {
  const key = requireKey(apiKey)
  const endpoint =
    url || process.env.AIGHT_EVENTS_INGEST_URL || DEFAULT_INGEST_URL.replace('/spans', '/events')
  const event = {
    event_id: eventId || randomUUID(),
    event_name: eventName,
    value,
    currency,
    // Unix *seconds*, float — not milliseconds. A ms timestamp lands ~50,000
    // years out and matches nothing.
    timestamp: timestamp ?? Date.now() / 1000,
    trace_id: traceId,
  }
  const res = await send(endpoint, key, event)
  return res.json()
}
